compiler/rustc_resolve/src/late.rs

```diff
@@ -1375,10 +1477,30 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
     }
 
     #[tracing::instrument(level = "debug", skip(self))]
-    fn create_fresh_lifetime(&mut self, id: NodeId, ident: Ident, item_node_id: NodeId) {
+    fn create_fresh_lifetime(
+        &mut self,
+        id: NodeId,
+        ident: Ident,
+        item_node_id: NodeId,
+    ) -> LifetimeRes {
         debug_assert_eq!(ident.name, kw::UnderscoreLifetime);
         debug!(?ident.span);
-        let item_def_id = self.r.local_def_id(item_node_id);
+        let item_def_id = self.r.opt_local_def_id(item_node_id).unwrap_or_else(|| {
+            // `item_node_id` does not always have an associated `LocalDefId`.  This can happen for
+            // `fn` types or parenthesized `Fn` traits.  In this case, look for the surrounding
+            // item and make it the parent of the new lifetime parameter's `LocalDefId`.
+            self.lifetime_ribs
+                .iter()
+                .rev()
+                .find_map(|rib| match rib.kind {
+                    LifetimeRibKind::Generics { binder, .. }
+                    | LifetimeRibKind::NoGenerics { binder } => self.r.opt_local_def_id(binder),
+                    _ => None,
```

Comment:
This list of `LifetimeRibKind`s should probably be explicit, with impossible cases (if any) producing panic.
I'm not sure what happens if we encounter e.g. `LifetimeRibKind::Item` on our way to the closest poly trait ref.